### PR TITLE
fea/infer memory optim

### PR DESCRIPTION
**Pull request by Superjomn**
Memory optimization pass for inference with pre-analysis of memory usage without GC.

It works with the following steps:

- Run the inference with several batches of fake data, important to try different sequence lengths if sequence data is used as inputs.
   - the predictor will collect the shapes of all the Tensors and persist to a cache file.
- Run the inference formally, the memory optimization pass will load the cache file, split the tensors into several clusters by the shape (currently just the batch size) change behaviors.
   - try to make a plan for reusing the memory of the existing tensors, attention that, a tensor can only reuse the ones in the same cluster.
      - the fixed shape tensor can reuse all the others with the fixed shape,
      - the tensor with LoD enabled can only reuse the others with the same sequence length,
      - the output of some `pool` op will change the first dimension to one(for each sequence), it can only share the memory with the other ones with the same behavior (TODO analysis all the shape dimensions).

It should work with both the LoD or Non-LoD models.

Currently, I've tested it with DAM, and that saves about 77.8% memory.

### CODE REVIEW

File: paddle/fluid/inference/tests/api/analyzer_text_classification_tester.cc [hunk 1]
```diff
@@ -106,6 +107,27 @@ TEST(Analyzer_Text_Classification, compare) {
       reinterpret_cast<const PaddlePredictor::Config *>(&cfg), input_slots_all);
 }
 
+TEST(Analyzer_Text_Classification, compare_with_memory_optim) {
+  AnalysisConfig cfg, cfg1;
+  // Run the first time to force to update memory cache
+  SetConfig(&cfg);
+  cfg.EnableMemoryOptim(true);
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/analyzer_text_classification_tester.cc]: - EnableMemoryOptim接口名要改为EnableMemoryOpt？或者EnableMemoryOptimzation么？
- 默认EnableMemoryOptim是开启的么？

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc [hunk 2]
```diff
@@ -218,5 +218,28 @@ TEST(Analyzer_dam, compare) {
...
+  cfg.EnableMemoryOptim(true);
+  cfg.Build();
+
+  std::vector<std::vector<PaddleTensor>> input_slots_all;
+  SetInput(&input_slots_all);
+
+  CompareNativeAndAnalysis(
+      reinterpret_cast<const PaddlePredictor::Config *>(&cfg), input_slots_all);
+
+  // Run second time to use the memory cache and perform memory optimaztion.
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/analyzer_dam_tester.cc]: - 请问在多线程下要怎么使用`EnableMemoryOptim`接口呢？
我在`TEST(Analyzer_dam, profile)`里加入：
```
  SetConfig(&cfg);
  cfg.EnableMemoryOptim(true);
  cfg.Build();
```
然后跑多线程，内存使用量没有变化。请问正确的使用方法是什么呢？
- > Currently, I've tested it with DAM, and that saves about 77.8% memory

请问要怎么复现这个结果呢？

File: paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+#include "paddle/fluid/string/pretty_log.h"
+
+namespace paddle {
+namespace inference {
+namespace analysis {
+
+using framework::ir::Graph;
+using framework::ir::Node;
+
+const int kBatchSize = 13;  // replace the placement -1 in shape
```
Comment (NHZlX) [inline paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc]: why 13 here?

File: paddle/fluid/inference/analysis/passes/memory_optimize_pass.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+#pragma once
+#include "paddle/fluid/inference/analysis/analysis_pass.h"
+#include "paddle/fluid/inference/analysis/passes/memory_optimize_pass.h"
+
+namespace paddle {
+namespace inference {
+namespace analysis {
+
+class MemoryOptimizePass : public AnalysisPass {
```
Comment (NHZlX) [inline paddle/fluid/inference/analysis/passes/memory_optimize_pass.h:48]: I feel that this class should be explained with more comments.

File: paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+      node->Op()->Flush();
+    }
+  }
+}
+
+void MemoryOptimizePass::PerformReusePlan(
+    const std::unordered_map<std::string, std::string>& reuse_table,
+    int sort_kind) const {
+  std::unordered_map<std::string, Node*> var_node_table;
+  BuildVarNodeTable(graph_, &var_node_table);
```
Comment (NHZlX) [inline paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:397]: This function looks useless.

File: paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+  free_existing_tensors->insert(free_tensor);
+}
+
+// Reuse a free existing tensor.
+void ReuseATensor(const std::string& tensor, const std::string& tensor2reuse,
+                  std::unordered_set<std::string>* free_existing_tensors,
+                  std::unordered_map<std::string, std::string>* reuse_table) {
+  auto it = free_existing_tensors->find(tensor2reuse);
+  PADDLE_ENFORCE(it != free_existing_tensors->end());
+  free_existing_tensors->erase(it);
```
Comment (NHZlX) [inline paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:221]: Why delete it from free_existing_tensors ?

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc (hunk 2 above)
Comment (Superjomn) [reply paddle/fluid/inference/tests/api/analyzer_dam_tester.cc]: 这是计算出来的值，现在profile 封装有问题，测试数据和vector input， vector output 也会在内存里，所以实际内存占用不一定准。

File: paddle/fluid/inference/tests/api/analyzer_text_classification_tester.cc (hunk 1 above)
Comment (Superjomn) [reply paddle/fluid/inference/tests/api/analyzer_text_classification_tester.cc]: 感觉 EnableMemoryOptim 应该差不多，不需要改

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc (hunk 2 above)
Comment (Superjomn) [reply paddle/fluid/inference/tests/api/analyzer_dam_tester.cc]: 71%是指定完复用策略，算出来的理论值， （未分配内存的tensor / 原始的所有tensor内存）
然后是temp variable的，parameter不算在里面。

File: paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc (hunk 3 above)
Comment (Superjomn) [reply paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc]: 一个特殊的数字来代表batch size，便于计算大概的内存消耗

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc (hunk 2 above)
Comment (Superjomn) [reply paddle/fluid/inference/tests/api/analyzer_dam_tester.cc]: 另外，现有的 buddy_allocator 可能也有问题，会额外占用内存，所以不一定能在实际内存中显示出来，需要更多修改。

File: paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc (hunk 6 above)
Comment (Superjomn) [reply paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:221]: the `free_existing_tensors` stores all the existing tensor that has been allocated memory and the memory is ready for reuse.
If an existing tensor is already reused and not free yet, it should be removed from this container variable.

**Pull request closed without merging**